drivers/volume/portworx/portworx.go

```diff
@@ -510,174 +514,1520 @@ func (d *portworx) RecoverDriver(n node.Node) error {
+				}
+			case api.SpecShared:
+				if requestedSpec.Shared != vol.Spec.Shared {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.Shared, vol.Spec.Shared)
+				}
+			case api.SpecSticky:
+				if requestedSpec.Sticky != vol.Spec.Sticky {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.Sticky, vol.Spec.Sticky)
+				}
+			case api.SpecGroup:
+				if !reflect.DeepEqual(requestedSpec.Group, vol.Spec.Group) {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.Group, vol.Spec.Group)
+				}
+			case api.SpecGroupEnforce:
+				if requestedSpec.GroupEnforced != vol.Spec.GroupEnforced {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.GroupEnforced, vol.Spec.GroupEnforced)
+				}
+			// portworx injects pvc name and namespace labels so response object won't be equal to request
+			case api.SpecLabels:
+				for requestedLabelKey, requestedLabelValue := range requestedLocator.VolumeLabels {
+					// check requested label is not in 'ignore' list
+					if labelValue, exists := vol.Locator.VolumeLabels[requestedLabelKey]; !exists || requestedLabelValue != labelValue {
+						return errFailedToInspectVolume(appVol, k, requestedLocator.VolumeLabels, vol.Locator.VolumeLabels)
+					}
+				}
+			case api.SpecIoProfile:
+				if requestedSpec.IoProfile != vol.Spec.IoProfile {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.IoProfile, vol.Spec.IoProfile)
+				}
+			case api.SpecSize:
+				if requestedSpec.Size != vol.Spec.Size {
+					return errFailedToInspectVolume(appVol, k, requestedSpec.Size, vol.Spec.Size)
+				}
+			default:
 			}
-		case api.SpecParent:
-			if v != vol.Source.Parent {
-				return errFailedToInspectVolume(volumeName, k, v, vol.Source.Parent)
+		}
+		err = d.ValidateVps(vol, appVols, pxNodes, nodeReplMap, appVolNodes)
+		if err != nil {
+			return &ErrFailedToInspectVolume{
+				ID:    appVol,
+				Cause: fmt.Sprintf("failed to validate VolumePlacementStratergy. Err: %v", err),
 			}
-		case api.SpecEphemeral:
-			if requestedSpec.Ephemeral != vol.Spec.Ephemeral {
-				return errFailedToInspectVolume(volumeName, k, requestedSpec.Ephemeral, vol.Spec.Ephemeral)
+		}
+		logrus.Infof("Successfully inspected volume: %v (%v)", vol.Locator.Name, vol.Id)
+
+	}
+
+	return nil
+}
+
+// Validate the volume replicas as per VolumePlacementStrategy rule applied to the volume
+func (d *portworx) ValidateVps(vol *api.Volume, appVols map[string]map[string]string, volNodes []api.StorageNode, nodeReplMap map[string][]*api.Volume, appVolNodes map[string]api.StorageNode) error {
+
+	logrus.Debugf("Volume details: %v ===\n (%v) Volumes per Node List:%v, all nodes of the App's Volumes: %v ", vol, appVols, nodeReplMap, appVolNodes)
+	logrus.Infof("Validate VPS  for Volume:%v ,  VPS Rule :(%v)", vol.Id, vol.Spec.GetPlacementStrategy())
+
+	if vpsrule := vol.Spec.GetPlacementStrategy(); vpsrule != nil {
```

Comment:
why do we need this check? Just check if placement_strategy is present in the volume labels. If yes, placement is enabled